### PR TITLE
Additional EWR features

### PR DESCRIPTION
**Pull request by emrahkocaman**
oss part of https://github.com/hazelcast/hazelcast-enterprise/issues/588

### CODE REVIEW

**Comment by emrahkocaman** [comment]
run-lab-run

**Comment by enesakar** [comment]
@pveentjer can you review this PR asap? we need this for EA3 release

File: hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java [hunk 1]
```diff
@@ -500,6 +480,46 @@ private void handleWanReplication(final Node node) throws Exception {
...
+            String queueFullBehavior = getTextContent(targetChild);
+            wanTarget.setQueueFullBehavior(WanQueueFullBehavior.valueOf(upperCaseInternal(queueFullBehavior)));
+        } else if ("batch-size".equals(targetChildName)) {
+            int batchSize = getIntegerValue("batch-size", getTextContent(targetChild));
+            wanTarget.setBatchSize(batchSize);
+        } else if ("batch-max-delay-millis".equals(targetChildName)) {
+            long batchMaxDelayMillis = getLongValue("batch-max-delay-millis", getTextContent(targetChild));
+            wanTarget.setBatchMaxDelayMillis(batchMaxDelayMillis);
+        } else if ("queue-capacity".equals(targetChildName)) {
+            int queueCapacity = getIntegerValue("queueCapacity", getTextContent(targetChild));
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: `getIntegerValue("queueCapacity"...` or getIntegerValue("queue-capacity"... ?

File: hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java [hunk 2]
```diff
@@ -500,6 +480,46 @@ private void handleWanReplication(final Node node) throws Exception {
         config.addWanReplicationConfig(wanReplicationConfig);
     }
 
+    private void handleWanTargetConfig(WanReplicationConfig wanReplicationConfig,
+                                       WanTargetClusterConfig wanTarget, Node targetChild) {
+        final String targetChildName = cleanNodeName(targetChild);
+        if ("replication-impl".equals(targetChildName)) {
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: what if there is no class, fail-fast? what do you think @mesutcelik ?

Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: we can set a default version....

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: Node start will fail with `ClassNotFoundException` in that case. But I think this part is out of scope of this since there is nothing new here. I just extracted them to a method logic is same as before.

File: hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    public int getId() {
+        return id;
+    }
+
+    /**
+     * Returns the {@link WanQueueFullBehavior} for the given id.
+     *
+     * @return the {@link WanQueueFullBehavior} found or null if not found
+     */
+    public static WanQueueFullBehavior getById(final int id) {
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java]: getById seems not used...

File: hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Determines the behavior of WAN replication impl. in case of WAN event queues are full
+ */
+public enum WanQueueFullBehavior {
+
+    /**
+     * Instruct WAN repl. impl to drop new events when WAN event queues are full
+     */
+    DISCARD_AFTER_MUTATION(0),
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java]: what does `DISCARD_AFTER_MUTATION` mean?

File: hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java (hunk 1 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: Should be `queue-capacity`

File: hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java [hunk 5]
```diff
@@ -24,6 +24,16 @@
 public class WanTargetClusterConfig {
 
     private static final WanAcknowledgeType DEFAULT_ACK_TYPE = WanAcknowledgeType.ACK_ON_OPERATION_COMPLETE;
+    private static final WanQueueFullBehavior QUEUE_FULL_BEHAVIOR = WanQueueFullBehavior.DISCARD_AFTER_MUTATION;
+    private static final int DEFAULT_BATCH_SIZE = 500;
+    private static final long DEFAULT_BATCH_MAX_DELAY_MILLIS = 1000;
+    private static final int DEFAULT_QUEUE_CAPACITY = 10000;
+    private static final long DEFAULT_RESPONSE_TIMEOUT_MILLIS = 60000;
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java]: I see those final statics not used anywhere else so what is the point to declare them here and then assing to class fields?

File: hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java (hunk 3 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java]: In fact, ID is totally unnecessary since this config is not serialized. I'll remove.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java [hunk 6]
```diff
@@ -68,6 +68,14 @@ public PutAllOperation(String name, MapEntries mapEntries, boolean initialLoad)
     }
 
     @Override
+    public void beforeRun() throws Exception {
+        super.beforeRun();
+        if (shouldWanReplicate()) {
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java]: What is the diff between `mapContainer.isWanReplicationEnabled()` and `shouldWanReplicate`?

File: hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java (hunk 4 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java]: Actually this is the existing behavior of WAN replication and will be default behavior of the new version also.
When `DISCARD_AFTER_MUTATION` is chosen, mutating operation (put/remove/etc...) will be applied but WAN event will be discarded if WAN replication queue is full.

When `THROW_EXCEPTION` is chosen, a queue size check will be performed in `beforeRun` of mutating operations and `WANReplicationQueueFullException` will be throwed if WAN replication queue is full, therefore mutating operation (put/remove/etc...) won't run.

File: hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java (hunk 5 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java]: Checkstyle doesn't allow magic numbers.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAndRemoveOperation.java [hunk 7]
```diff
@@ -37,6 +37,13 @@ public CacheGetAndRemoveOperation(String name, Data key, int completionId) {
     }
 
     @Override
+    protected void beforeRunInternal() {
+        if (cache.isWanReplicationEnabled()) {
```
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAndRemoveOperation.java]: not important but `super.beforeRunInternal();` is needed here also?

File: hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java (hunk 4 above)
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/WanQueueFullBehavior.java]: ok

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/AbstractMultipleEntryOperation.java [hunk 8]
```diff
@@ -67,6 +67,15 @@ protected AbstractMultipleEntryOperation(String name, EntryBackupProcessor backu
         this.backupProcessor = backupProcessor;
     }
 
+
+    @Override
+    public void beforeRun() throws Exception {
+        super.beforeRun();
+        if (mapContainer.isWanReplicationEnabled()) {
+            mapContainer.checkWanReplicationQueues();
+        }
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/AbstractMultipleEntryOperation.java]: We have a [MapOperation] (https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java#L52) to centralize this kind of checks.

File: hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java (hunk 5 above)
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/WanTargetClusterConfig.java]: ok. I would prefer to create a constructor which is accepting all class fields and remove all setters. Default constructor would use those static fields. However, that is my preference...

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java (hunk 6 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java]: Seems like they are same. It's possibly a leftover from times before `isWanReplicationEnabled` method is introduced.

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java]: I'll remove `shouldWanReplicate` method.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAndRemoveOperation.java (hunk 7 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAndRemoveOperation.java]: :+1: I think so.

File: hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java (hunk 2 above)
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/XmlConfigBuilder.java]: I think that is generic problem with all configs that accept fully qualified class names... @bilalyasar , can you please create an issue for that?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/AbstractMultipleEntryOperation.java (hunk 8 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/AbstractMultipleEntryOperation.java]: But we need to make this check only in mutating operations. Otherwise, for example a get operation  will also throw exception when WAN queue is full.

**Comment by emrahkocaman** [comment]
I've updated the PR.

For map, @ahmetmircik, I've moved WAN checks to `WANAwareOperationProvider` and cleaned `beforeRun` methods.

For cache, @serkan-ozal , Also cleaned up WAN related checks from cache operations. They'll be available in ee.

Also I've added tests for configuration part.

Guys, can you please check again?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviders.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    private final MapOperationProvider defaultProvider = new DefaultMapOperationProvider();
+
+    public MapOperationProviders(MapServiceContext mapServiceContext) {
+        this.mapServiceContext = mapServiceContext;
+        this.wanAwareProvider = new WANAwareOperationProvider(mapServiceContext, defaultProvider);
+    }
+
+    /**
+     * Creates {@link MapOperationProvider} instance and wraps it into a {@link WANAwareOperationProvider} if
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviders.java]: `if` seems leftover in the end.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviderDelegator.java [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ * Base class which basically delegates all method calls to underlying {@link MapOperationProvider}
+ * to prevent extending classes to override all methods.
+ *
+ * See {@link WANAwareOperationProvider} for an example.
+ */
+public abstract class MapOperationProviderDelegator implements MapOperationProvider {
+
+    abstract MapOperationProvider getOperationProviderDelegate();
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviderDelegator.java]: Maybe renamed as `getDelegate()`, we know this is map-operation-provider from context.

**Comment by ahmetmircik** [comment]
:+1:  with minor comments.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviders.java (hunk 9 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviders.java]: :+1:

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviderDelegator.java (hunk 10 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperationProviderDelegator.java]: I'll update.

**Comment by emrahkocaman** [comment]
run-lab-run

**Comment by mesutcelik** [comment]
:+1:
